beancount-core/src/date.rs

```diff
@@ -0,0 +1,47 @@
+use std::borrow::Cow;
+
+#[cfg(feature = "chrono")]
+use chrono::NaiveDate;
+
+#[derive(Eq, PartialEq, Debug, Clone)]
+pub struct Date<'a>(Cow<'a, str>);
+
+impl<'a> From<Cow<'a, str>> for Date<'a> {
+    fn from(s: Cow<'a, str>) -> Self {
+        Date(s)
+    }
+}
+
+impl<'a> From<&'a str> for Date<'a> {
+    fn from(s: &'a str) -> Self {
+        Date(Cow::from(s))
+    }
+}
+
+impl From<String> for Date<'static> {
+    fn from(s: String) -> Self {
+        Date(Cow::from(s))
+    }
+}
+
+impl<'a> From<Date<'a>> for Cow<'a, str> {
+    fn from(d: Date<'a>) -> Self {
+        d.0
+    }
+}
+
+#[cfg(feature = "chrono")]
+impl From<NaiveDate> for Date<'static> {
```

Comment:
Can this be the anonymous lifetime instead (e.g. `Date<'_>`) rather than the static lifetime?  Or is there a reason the static lifetime was used here?  I believe this change would allow users to use convert dates of any lifetime, rather than only those with a static lifetime.

Same question for the other two static lifetimes in this PR.